support HRA

DaShenZi721 opened the pull request:
Hello! I'm the author of [Bridging The Gap between Low-rank and Orthogonal Adaptation via Householder Reflection Adaptation](https://arxiv.org/abs/2405.17484). 
This PR integrates our method HRA.

BenjaminBossan reviewed (changes requested):
Thanks a lot for this PR to add HRA to PEFT. This method looks very promising and I'm happy to add it to PEFT.

I took a look at the paper (which I have to admit was hard for me to understand) and the code, which already looks quite good.

I didn't start the CI yet, because there is currently an issue with PyTorch and numpy 2.0 which is hopefully resolved very soon. However, I ran some tests locally and found a few of them failing. Could you please check if you find the same when you run `pytest tests/ -k hra` locally?

I haven't checked all tests but it appears that at least some of them are related to the initialization. By default, you have `init_weights=True`, which leads to Kaiming uniform initialization. As a result, with an untrained HRA adapter, calling forward is _not_ an identity transform. Is that understanding correct?

To get an identity transform, presumably we have to set `init_weights=False`. Note that this is the opposite of what we have for the other adapters like LoRA: There, if we set `init_weights=True` or `init_lora_weights=True`, it is the identity transform, i.e. at the start of the training, the adapter has no effect on the output.

There can be three solutions to this:

1. You change the default to `init_weights=False`, but this can be confusing to users, as the default for the other methods is `True`. 
2. You leave the default as is but for the failing tests, you have to set `init_weights=False`.
3. You change the meaning such that `init_weights=True` leads to zero-initialization. This is more in line with what LoRA etc. do but probably does not conform to the paper.

Note that I tested setting `init_weights=False` to see if this leads to the identity transform but instead `forward`  returned `nan`. So this needs to be checked too.

Before I do a full in-depth review of the code, could you please also do the following:

1. Add some documentation. It would be especially helpful to write a short, easy to digest description of the method and how it compares to other methods, especially LoRA/OFT/BOFT. Also, let's try to give good description of the hyper parameters so that users know how to tweak them without having to read and understand the whole paper.
2. It would be really nice if you could add an example to the `examples/` folder. Maybe something based of the original repo that we can use to check that the PEFT implementation is faithful.
3. Speaking of the original repo, if some of the code in this PR is a 1:1 copy of https://github.com/DaShenZi721/HRA, it can be helpful to mark this with code comments and perma-links.
4. Please run `make style` on the PR.

DaShenZi721 commented:
Hello, thank you for your suggestions! Here are the updates on the PR:

1. Modified the initialization. Currently, if we set `init_weights=True`, HRA will be the identity transform, i.e. at the start of the training, the adapter has no effect on the output.
2. Fixed some bugs, and now all test cases pass!
3. Added an example: `hra_dreambooth`. We follow the examples of oft and boft. Additionally, we have included a simple README. You can try it by following the instructions provided in the README!
4. Ran `make style`.

Thank you for your review! If there are any issues, please let me know.

BenjaminBossan commented:
Thanks @DaShenZi721 for the updates. Could you please run `make style` on the PR? Note that we updated the ruff version, so in your env, you may have to run `python -m pip install -U ruff` first.

DaShenZi721 commented:
OK, I just updated the ruff version from 0.2.2 to 0.4.10, and ran `make style` to modify the two files.

BenjaminBossan reviewed (changes requested):
Thanks a lot for the additions to the PR. It is much more complete now.

While reviewing, I found a few smaller issues. There is also a potentially bigger one, namely in the forward methods the layers. Please check my comments, where I described the details.

Also thanks for adding the examples. I haven't reviewed them yet, as I think we must first address the other issues. I'll get to that once the rest is sorted out.

DaShenZi721 commented on tests/test_custom_models.py:
```diff
@@ -245,6 +246,14 @@
+    ########
+    ("Vanilla MLP 1 HRA", "MLP", HRAConfig, {"target_modules": "lin0"}),
+    ("Vanilla MLP 2 HRA", "MLP", HRAConfig, {"target_modules": ["lin0"]}),
+    ("Vanilla MLP 3 HRA", "MLP", HRAConfig, {"target_modules": ["lin0", "lin1"]}),
```
Gram-Schmidt orthogonalization results in strict orthogonality, which prevents an untrained HRA from being initialized as an identity transform. Consequently, many test cases related to initialization would fail. Therefore, I did not add a test with `apply_GS=True`.

BenjaminBossan replied:
I see. Still, it would be nice if we could test that, to ensure that this part of the code is well covered. I have a suggestion: Maybe you could add a test to `tests_common_gpu.py` that is similar to this one:

https://github.com/huggingface/peft/blob/7ffa43b16e4625322c9f157db12c9358f2f54cdf/tests/test_common_gpu.py#L946

(but bnb quantization is not needed).

The test could instantiate a causal LM, create the output, then apply PEFT with HRA+GS and check the output again. The two outputs should be different.

If there is something else that could be tested, e.g. by checking that some weights are initialized in a certain way, that could also be added. WDYT?

DaShenZi721 replied:
OK, we have added a test case for HRA+GS, as well as a test case to verify that HRA cannot be initialized as an identity transform when r is an odd number.

DaShenZi721 commented on src/peft/tuners/hra/layer.py:
```diff
@@ -0,0 +1,440 @@
+                delta_weight = self.get_delta_weight(active_adapter)
+                new_weight = torch.mm(orig_weight, delta_weight)
+
+                result = F.linear(input=x, weight=new_weight, bias=self.base_layer.bias)
```
Sorry, we didn't account for the multiple activate adapters. This bug has now been fixed, and we have added and successfully passed the test cases you mentioned.

DaShenZi721 commented on src/peft/tuners/hra/config.py:
```diff
@@ -0,0 +1,112 @@
+            "help": "The layer pattern name, used only if `layers_to_transform` is different to None and if the layer pattern is not in the common layers pattern."
+        },
+    )
+    fan_in_fan_out: bool = field(
```
Done.

DaShenZi721 commented on src/peft/tuners/hra/config.py:
```diff
@@ -0,0 +1,112 @@
+            Whether to apply Gram-Schmidt orthogonalization.
+    """
+
+    r: int = field(default=8, metadata={"help": "HRA rank"})
```
Done.

DaShenZi721 commented on src/peft/tuners/hra/__init__.py:
```diff
@@ -0,0 +1,20 @@
+# Copyright 2023-present the HuggingFace Inc. team.
```
Done.

DaShenZi721 commented on src/peft/tuners/hra/layer.py:
```diff
@@ -0,0 +1,440 @@
+                    unit_v = opt_u[:, i].view(-1, 1)
+                    weight = weight @ (
+                        torch.eye(shape[0], device=opt_u.device, dtype=opt_u.dtype) - 2 * unit_v @ unit_v.t()
+                    )
```
Done.

BenjaminBossan commented on README.md:
```diff
@@ -1,5 +1,5 @@
 <!---
-Copyright 2023 The HuggingFace Team. All rights reserved.
+Copyright 2024 The HuggingFace Team. All rights reserved.
```
It seems that you have replaced all the years in all files. This should not be done. Please revert this change. The year only needs to be updated in the files that you created for this PR.

DaShenZi721 replied:
I'm sorry...

DaShenZi721 commented:
Hi! Do you find any other problems?

BenjaminBossan commented on src/peft/tuners/hra/model.py:
```diff
@@ -0,0 +1,335 @@
+    @staticmethod
+    def _prepare_adapter_config(peft_config, model_config):
+        if peft_config.target_modules is None:
+            if model_config["model_type"] not in TRANSFORMERS_MODELS_TO_LORA_TARGET_MODULES_MAPPING:
```
You're re-using the LoRA mapping here (which I realize BOFT also does). I guess it's okay, although there could be some LoRA models that will not work with HRA.

DaShenZi721 replied:
Yes, because HRA with the LoRA mapping has passed all the tests and BOFT does it this way as well, we decided to reuse it.

BenjaminBossan reviewed (changes requested):
Thanks for the updates, good work. Overall, this looks pretty good and I only found a few minor issues that need to be addressed before we can merge -- please check my comments. Apart from those, could you please run `make style`?

DaShenZi721 commented on examples/hra_dreambooth/README.md:
```diff
@@ -0,0 +1,112 @@
+<!--Copyright 2023 The HuggingFace Team. All rights reserved.
```
Done.

DaShenZi721 commented on src/peft/tuners/hra/layer.py:
```diff
@@ -0,0 +1,443 @@
+            if active_adapter not in self.hra_u.keys():
+                continue
+
+            warnings.warn("Unscaling operation for HRA not supported! Keeping scale to 1.")
```
Done.

DaShenZi721 commented on examples/hra_dreambooth/README.md:
```diff
@@ -0,0 +1,112 @@
+Therefore, r is usually set to 8.
+**Note**, please set r to an even number to avoid potential issues during initialization.
+- `hra_apply_GS`: Applys Gram-Schmidt orthogonalization. Default is `false`.
+- `hra_bias`: specify if the `bias` paramteres should be traind. Can be `none`, `all` or `hra_only`.
```
Done.

DaShenZi721 commented on examples/hra_dreambooth/README.md:
```diff
@@ -0,0 +1,112 @@
+HRA provides a new perspective connecting LoRA to OFT and achieves encouraging performance in various downstream tasks.
+HRA adapts a pre-trained model by multiplying each frozen weight matrix with a chain of r learnable Householder reflections (HRs).
+HRA can be interpreted as either an OFT adapter or an adaptive LoRA. 
+Consequently, it harnesses the advantages of both strategies, reducing parameters and computation costs while penalizing the loss of pre-training knowledge
```
Done.

DaShenZi721 commented on examples/hra_dreambooth/README.md:
```diff
@@ -0,0 +1,112 @@
+HRA adapts a pre-trained model by multiplying each frozen weight matrix with a chain of r learnable Householder reflections (HRs).
+HRA can be interpreted as either an OFT adapter or an adaptive LoRA. 
+Consequently, it harnesses the advantages of both strategies, reducing parameters and computation costs while penalizing the loss of pre-training knowledge
+For further details on BOFT, please consult the [original HRA paper](https://arxiv.org/abs/2405.17484).
```
Done.

DaShenZi721 commented on tests/test_torch_compile.py:
```diff
@@ -77,6 +78,7 @@
     "lora-with-modules-to-save": (LoraConfig(task_type=TaskType.CAUSAL_LM, modules_to_save=["embed_tokens"]), {}),
     "oft": (OFTConfig(task_type=TaskType.CAUSAL_LM, target_modules=["q_proj", "v_proj"]), {}),
     "vera": (VeraConfig(task_type=TaskType.CAUSAL_LM), {}),
+    "hra": (HRAConfig(task_type=TaskType.CAUSAL_LM, target_modules=["q_proj", "v_proj"]), {}),
```
Done.

DaShenZi721 commented on tests/test_common_gpu.py:
```diff
@@ -1083,6 +1084,54 @@ def test_8bit_dora_merging(self):
 
+    @require_torch_gpu
+    @pytest.mark.single_gpu_tests
+    def test_apply_GS_hra_inference(self):
```
OK, we have added a test case `test_apply_GS_hra_conv2d_inference()` for `apply_GS=True` and `Conv2d`, as well as a test case in `test_vision_models.py`.

DaShenZi721 commented on examples/hra_dreambooth/train_dreambooth.sh:
```diff
@@ -0,0 +1,190 @@
+    --class_data_dir="$CLASS_DIR" \
+    --output_dir="$OUTPUT_DIR" \
+    --instance_prompt="$instance_prompt" \
+    --with_prior_preservation --prior_loss_weight=1.0 \
```
Done.

DaShenZi721 commented on examples/hra_dreambooth/README.md:
````diff
@@ -0,0 +1,112 @@
+## Fine-tune Dreambooth with HRA
+
+```bash
+./train_dreambooth.sh $prompt_idx $class_idx
````
Change to `bash ./train_dreambooth.sh $prompt_idx $class_idx`

DaShenZi721 commented on examples/hra_dreambooth/train_dreambooth.py:
```diff
@@ -0,0 +1,1083 @@
+import argparse
```
Done.

DaShenZi721 commented on examples/hra_dreambooth/README.md:
```diff
@@ -0,0 +1,112 @@
+However, this also results in higher memory consumption and longer computation times. 
+Therefore, r is usually set to 8.
+**Note**, please set r to an even number to avoid potential issues during initialization.
+- `hra_apply_GS`: Applys Gram-Schmidt orthogonalization. Default is `false`.
```
Done.

BenjaminBossan commented on tests/test_stablediffusion.py:
```diff
@@ -1,4 +1,4 @@
-# Copyright 2023-present the HuggingFace Inc. team.
+# Copyright 2024-present the HuggingFace Inc. team.
```
Please undo.

DaShenZi721 replied:
Done.

BenjaminBossan commented on tests/test_common_gpu.py:
```diff
@@ -1083,6 +1084,84 @@ def test_8bit_dora_merging(self):
+    @pytest.mark.single_gpu_tests
+    def test_apply_GS_hra_conv2d_inference(self):
+        # check for different result with and without apply_GS
+        from datasets import load_dataset
+        from transformers import AutoImageProcessor, AutoModelForImageClassification
```
These imports can be on the root, not necessary to make them local.

BenjaminBossan commented on examples/hra_dreambooth/train_dreambooth.py:
```diff
@@ -0,0 +1,1083 @@
+
+if __name__ == "__main__":
+    args = parse_args()
+    main(args)
```
Indeed, I had to delete the `logs` directory and then after retraining, it works. Would be good to a add a comment about this for users who encounter the same problem -- or even better fixing the problem :)

BenjaminBossan commented on examples/hra_dreambooth/train_dreambooth.sh:
```diff
@@ -0,0 +1,191 @@
+    --hra_r=$hra_r 
+
+  echo "Please fill in the following parameters in the hra_dreambooth_inference.ipynb"
+  echo "BASE_MODEL_NAME = $MODEL_NAME"
+  echo "ADAPTER_MODEL_PATH = $OUTPUT_DIR"
+  echo "PROMPT = $instance_prompt"
```
Just a suggestion, this way users can copy&paste into the notebook without having to edit the code.

```suggestion
  echo "BASE_MODEL_NAME = '$MODEL_NAME'"
  echo "ADAPTER_MODEL_PATH = '$OUTPUT_DIR'"
  echo "PROMPT = '$instance_prompt'"
```

DaShenZi721 replied:
Done.

BenjaminBossan reviewed (changes requested):
This PR looks great, we're almost there. Just some very small comments from my side, please take a look.

BenjaminBossan commented on src/peft/tuners/hra/model.py:
```diff
@@ -0,0 +1,335 @@
+    def __getattr__(self, name: str):
+        """Forward missing attributes to the wrapped module."""
+        try:
+            return super().__getattr__(name)  # defer to nn.Module's logic
```
Just a small thing, we just merged a small bug in `__getattr__`: #1892. Could you please apply the fix here too?

BenjaminBossan replied:
There is also a merge conflict in `test_common_gpu.py`, which should hopefully be easy to resolve.

DaShenZi721 commented:
We have applied the fix #1892 and resolved the conflicts! 
Besides, we updated the hra_drembooth to generate higher quality images!

BenjaminBossan commented:
@DaShenZi721 Thanks for the updates. We just merged another PR, which resulted in many merge conflicts, but they should be very straightforward to resolve. Could you please take a look?

I'll depart soon to the EuroPython conference, so unfortunately, I'll not be able to review the PR this week.

DaShenZi721 commented:
We have resolved the conflicts.

BenjaminBossan commented on examples/hra_dreambooth/train_dreambooth.py:
```diff
@@ -0,0 +1,608 @@
+#!/usr/bin/env python
+# Copyright 2023-present the HuggingFace Inc. team.
```
```suggestion
# Copyright 2024-present the HuggingFace Inc. team.
```

DaShenZi721 replied:
Done.

BenjaminBossan commented on examples/hra_dreambooth/train_dreambooth.py:
```diff
@@ -0,0 +1,608 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# The implementation is based on "Parameter-Efficient Orthogonal Finetuning
```
Remove

DaShenZi721 replied:
Done.

BenjaminBossan commented on examples/hra_dreambooth/train_dreambooth.py:
```diff
@@ -0,0 +1,608 @@
+            }
+        }
+
+    # Currently, it's not possible to do gradient accumulation when training two models with accelerate.accumulate
```
Is this comment relevant?

DaShenZi721 replied:
Actually, this script is mainly adapted from `boft_dreambooth/train_dreambooth.py`, and we only modify the adapater part of this code. 
As long as we can use the HRA and generate images as normal, we don't care about the rest of the code.

BenjaminBossan commented on examples/hra_dreambooth/train_dreambooth.py:
```diff
@@ -0,0 +1,608 @@
+                # Checks if the accelerator has performed an optimization step behind the scenes
+                if accelerator.sync_gradients:
+                    progress_bar.update(1)
+                    # if args.report_to == "wandb":
```
Remove?

DaShenZi721 replied:
Done.

BenjaminBossan commented on examples/hra_dreambooth/train_dreambooth.py:
```diff
@@ -0,0 +1,608 @@
+                        generator = torch.Generator(device=accelerator.device).manual_seed(args.seed)
+                    else:
+                        generator = None
+                    # images = []
```
Remove?

DaShenZi721 replied:
Done.

BenjaminBossan commented on examples/hra_dreambooth/utils/args_loader.py:
```diff
@@ -0,0 +1,363 @@
+        type=str,
+        default="wandb",
+        help=(
+            'The integration to report the results and logs to. Supported platforms are `"tensorboard"`'
+            ' (default), `"wandb"` and `"comet_ml"`. Use `"all"` to report to all integrations.'
```
Help says default is tensorboard, but it is wandb.

DaShenZi721 replied:
The default has been modified.

BenjaminBossan commented on examples/hra_dreambooth/train_dreambooth.py:
```diff
@@ -0,0 +1,608 @@
+    accelerator = Accelerator(
+        gradient_accumulation_steps=args.gradient_accumulation_steps,
+        mixed_precision=args.mixed_precision,
+        log_with=args.report_to,
```
Currently, there is no option to disable logging completely, is there? How about

```suggestion
        log_with=args.report_to if args.report_to != "none" else None,
```

If this is changed, please adapt the help of the argument accordingly.

DaShenZi721 replied:
Done.

BenjaminBossan commented on examples/hra_dreambooth/utils/args_loader.py:
```diff
@@ -0,0 +1,363 @@
+import argparse
```
The new utils files are basically copied from the existing ones, right? Let's add a comment to the files: `# copied from ...`  or `# adapted from ...`.

DaShenZi721 replied:
Done.

BenjaminBossan commented on examples/hra_dreambooth/train_dreambooth.py:
```diff
@@ -0,0 +1,608 @@
+            target_modules=UNET_TARGET_MODULES,
+            bias=args.hra_bias,
+        )
+        unet = get_peft_model(unet, config, adapter_name=args.wandb_run_name)
+        unet.print_trainable_parameters()
+
+    vae.requires_grad_(False)
+    unet.train()
+
+    if args.train_text_encoder and args.use_hra:
+        config = HRAConfig(
+            r=args.hra_r,
+            apply_GS=args.hra_apply_GS,
+            target_modules=UNET_TARGET_MODULES,
+            bias=args.hra_bias,
+        )
+        text_encoder = get_peft_model(text_encoder, config, adapter_name=args.wandb_run_name)
```
Hmm, not a fan of using `args.wandb_run_name` here, what if users don't want to use wandb? Maybe the argument could be renamed `run_name`?

DaShenZi721 replied:
Done. This script, which previously had to use wandb as the logger, is no longer needed now.
You could easily switch the logger or set no logger at all.

BenjaminBossan commented on examples/hra_dreambooth/train_dreambooth.py:
```diff
@@ -0,0 +1,608 @@
+    # We need to initialize the trackers we use, and also store our configuration.
+    # The trackers initializes automatically on the main process.
+    if accelerator.is_main_process:
+        accelerator.init_trackers(args.wandb_project_name, config=vars(args), init_kwargs=wandb_init)
```
This fails if not using wandb.

DaShenZi721 replied:
Done.

BenjaminBossan reviewed (changes requested):
The code and the recent changes all look good. I had a fresh look at the training script and actually found a couple of issues when trying not to use wandb for logging. Maybe these issues are already in the original code, not sure, but please take a look. It would be nice if users could easily switch the logger or set no logger at all. Besides this, the PR is good to go.

DaShenZi721 commented on examples/hra_dreambooth/README.md:
````diff
@@ -0,0 +1,98 @@
+
+```
+hra_dreambooth
+├── data
````
The `data` directory should have been automatically created when running the `train_dreambooth.sh`.

BenjaminBossan reviewed (approved):
Thanks for fixing the training script, it now works well for me without wandb. And also big thanks for adding HRA to PEFT. I'll merge the PR once CI is green.

DaShenZi721 commented:
Thank you so much for your careful and responsible guidance!

BenjaminBossan commented:
Hi @DaShenZi721, I noticed that one of our unit tests was particularly slow and it is involves HRA. The test is `test_model_with_batchnorm_reproducibility` in `tests/test_vision_models.py`. On [CI](https://github.com/huggingface/peft/actions/runs/10720719938/job/29732290034#step:6:505), this test takes > 500 sec. But even locally on GPU, it takes > 70 sec for me. I did some quick tests and it's the forward and backwards passes that take a lot of time. The same test with LoRA just takes 7 sec locally.

I wonder if you have some ideas what the exact reason could be and if we could speed up HRA, which would also help with adoption. Otherwise, I'll probably just remove this test.

DaShenZi721 commented:
Sorry, I missed this message... I will now look into how to address this issue.

DaShenZi721 commented:
I wasn't able to reproduce the issue. Has it already been resolved?

BenjaminBossan commented:
> I wasn't able to reproduce the issue. Has it already been resolved?

I [switched the underlying model](https://github.com/huggingface/peft/commit/54be5a3db61748d698ca2e6b55bcfef229a9b475) which is probably why you can't reproduce now. From that point of view, this is no longer urgent. I think it could still be worth investigating why `microsoft/resnet-18` was so slow.

DaShenZi721 commented:
OK, I'll check it out.
